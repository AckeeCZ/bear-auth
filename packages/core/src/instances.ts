import type { BearAuth } from './create';
import { BearAuthError } from './errors';

export const instances = new Map<string, BearAuth<unknown>>();

export function getInstance<AuthInfo>(instanceId: BearAuth<AuthInfo>['id']) {
    const instance = instances.get(instanceId);

    if (!instance) {
        throw new BearAuthError(
            'bear-auth/unknown-instance',
            `No BearAuth instance found for ${instanceId} ID. Call 'create(instanceId: string)' first.`,
        );
    }

    return instance as BearAuth<AuthInfo>;
}

export function setInstance<AuthState>(bearAuth: BearAuth<AuthState>) {
    instances.set(bearAuth.id, bearAuth as BearAuth<unknown>);
}

/**
 * Get all active BearAuth instance IDs.
 */
export function getCreatedBearAuths() {
    return Array.from(instances.keys());
}
