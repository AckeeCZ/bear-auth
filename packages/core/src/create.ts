import { BearAuthError } from '~/errors';
import { createDefaultLogger, defaultLogLevel, type Logger, type LogLevel } from '~/logger';
import type { StorageSchema } from '~/storage';
import { createInitialState, type State } from '~/store/state';

import { instances } from './instances';
import { defaultContinueWhenOnline } from './network';
import type { OnAuthStateChangedCallback } from './onAuthStateChanged';
import type { AuthenticatedSession, RefreshingSession, Session } from './store/session';

export type BearAuth<AuthInfo> = {
    id: string;

    /**
     * It's null if the custom logger is set.
     * @default 'error'
     */
    loglevel: LogLevel | null;
    logger: Logger;

    hooks: {
        refreshToken: null | ((authSession?: RefreshingSession<AuthInfo>['data']) => Promise<Session<AuthInfo>>);
        fetchAuthInfo: null | ((authSession?: AuthenticatedSession<AuthInfo>['data']) => Promise<Session<AuthInfo>>);
        logout: null | (() => Promise<void>);
    };
    state: State<AuthInfo>;
    flags: {
        autoRefreshAccessTokenEnabled: boolean;
        customLogger: boolean;
    };

    storage: StorageSchema<AuthInfo> | null;

    /**
     * Internal storage version. If increased, then during `start` method, the storage will be cleared.
     */
    storageVersion: 1;

    onAuthStateChanged: Set<OnAuthStateChangedCallback<AuthInfo>>;

    refreshTokenTimeoutId: null | (number | NodeJS.Timeout);

    continueWhenOnline: () => Promise<void>;
};

export interface CreateProps {
    /**
     * Choose a unique instance ID when creating a multiple BearAuth instances.
     */
    id?: string;
}

/**
 * Initialize BearAuth instance. Can't be called multiple times with the same instance ID (Call `destroy` first).
 * @returns a string reference to the created BearAuth instance.
 */
export function create({ id = 'bear_auth' }: CreateProps = {}) {
    if (instances.has(id)) {
        throw new BearAuthError(
            'bear-auth/unique-instance',
            `BearAuth with '${id}' ID has been already created. Call 'await detroy(bearAuth: BearAuth)' first.`,
        );
    }

    const instance = {
        id,

        loglevel: defaultLogLevel,
        logger: createDefaultLogger(id, defaultLogLevel),

        hooks: {
            refreshToken: null,
            fetchAuthInfo: null,
            logout: null,
        },

        state: createInitialState<unknown>(),

        flags: {
            autoRefreshAccessTokenEnabled: false,
            customLogger: false,
        },

        storage: null,

        storageVersion: 1,

        onAuthStateChanged: new Set<OnAuthStateChangedCallback<unknown>>(),

        refreshTokenTimeoutId: null,

        continueWhenOnline: defaultContinueWhenOnline,
    } as const satisfies BearAuth<unknown>;

    instances.set(id, instance);

    return instance.id;
}
