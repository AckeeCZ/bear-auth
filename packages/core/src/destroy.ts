import { stopTokenAutoRefresh } from './autoRefreshToken.ts';
import { type BearAuth } from './create.ts';
import { getInstance, instances } from './instances.ts';
import { runOnAuthStateChangedCallbacks } from './onAuthStateChanged.ts';
import { setUnauthenticatedSession } from './store/session.ts';
import { allTasksSettled, deleteInstanceTasks } from './tasks.ts';

/**
 * Destroys the BearAuth instance.
 * - Tries to logout user.
 * - Clears the storage.
 * - Deletes the instance.
 * @param id - return value of `create` method
 */
export async function destroy<AuthInfo>(id: BearAuth<AuthInfo>['id']) {
    const instance = getInstance<AuthInfo>(id);

    if (!instance) {
        return;
    }

    await stopTokenAutoRefresh<AuthInfo>(id);

    const { logger, authSessionPropagation, hooks, storage, store } = instance;

    logger.debug('[destroy]', 'Destroying auth session.');

    authSessionPropagation?.cleanUp();

    if (hooks.logout) {
        await hooks.logout();
    } else {
        await storage?.clear(id);

        await store.setSession(setUnauthenticatedSession);

        await runOnAuthStateChangedCallbacks<AuthInfo>(id);
    }

    instance.onAuthStateChanged.callbacks.clear();

    await allTasksSettled(id);
    deleteInstanceTasks(id);

    const keys = Object.keys(instance) as (keyof BearAuth<AuthInfo>)[];

    for (const key of keys) {
        delete instance[key];
    }

    instances.delete(id);
}
