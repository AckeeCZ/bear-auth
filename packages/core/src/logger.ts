import { getLogger } from 'loglevel';

import { type BearAuth } from '~/create';

import { BearAuthError } from './errors';
import { getInstance } from './instances';

export type LogLevel = 'error' | 'debug' | 'info' | 'silent';

export const defaultLogLevel: LogLevel = 'error';

export function createDefaultLogger(id: BearAuth<unknown>['id'], loglevel: LogLevel) {
    const { error, debug, info, setLevel } = getLogger(id);

    const defaultLogger = {
        error<Args extends unknown[]>(...args: Args) {
            error('[bear-auth]', ...args);
        },
        debug<Args extends unknown[]>(...args: Args) {
            debug('[bear-auth]', ...args);
        },
        info<Args extends unknown[]>(...args: Args) {
            info('[bear-auth]', ...args);
        },
    } as const;

    setLevel(loglevel);

    return defaultLogger;
}

export type Logger = ReturnType<typeof createDefaultLogger>;

/**
 * Customize the log level.
 * @param id - return value of `create` method
 */
export function setLogLevel(id: BearAuth<unknown>['id'], level: LogLevel) {
    const instance = getInstance(id);

    if (instance.flags.customLogger) {
        throw new BearAuthError('bear-auth/log-level', 'Cannot set log level when using custom logger.');
    }

    instance.loglevel = level;

    instance.logger.debug('[setLogLevel]', 'Setting log level', level);

    getLogger(id).setLevel(level);
}

/**
 * Override the default logger.
 * @param id - return value of `create` method
 * @param logger - custom logger
 */
export function setLogger<L extends Logger>(id: BearAuth<unknown>['id'], logger: L) {
    const instance = getInstance(id);

    const { debug, info, error } = logger;

    instance.logger = {
        debug,
        info,
        error,
    } as const;

    instance.flags.customLogger = true;
    instance.loglevel = null;

    debug('setLogger', 'Custom logger is set. The setLogLevel method will not do anything.');
}
