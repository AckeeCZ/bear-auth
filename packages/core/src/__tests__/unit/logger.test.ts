import { describe, expect, test, vi } from 'vitest';

import { create } from '../../create.ts';
import { destroy } from '../../destroy.ts';
import { instances } from '../../instances.ts';
import { defaultLogLevel, setLogger, setLogLevel, type Logger } from '../../logger.ts';

describe('logger', () => {
    test('should set log level', async () => {
        const id = 'test-instance-id';
        create({ id });

        expect(instances.get(id)?.loglevel).toBe(defaultLogLevel);

        setLogLevel(id, 'debug');

        expect(instances.get(id)?.loglevel).toBe('debug');

        await destroy(id);
    });

    test('should set custom logger', async () => {
        const id = 'test-instance-id';
        create({ id });

        const error = vi.fn();
        const debug = vi.fn();
        const info = vi.fn();

        const customLogger = {
            error(...args) {
                error(...args);
            },
            debug(...args) {
                debug(...args);
            },
            info(...args) {
                info(...args);
            },
        } satisfies Logger;

        setLogger(id, customLogger);

        const instance = instances.get(id);
        expect(instance?.logger).toEqual(customLogger);
        expect(instance?.flags.customLogger).toBe(true);
        expect(instance?.loglevel).toBe(null);

        expect(() => setLogLevel(id, 'debug')).toThrowErrorMatchingSnapshot(
            `cannot set log level when using custom logger`,
        );
    });
});
