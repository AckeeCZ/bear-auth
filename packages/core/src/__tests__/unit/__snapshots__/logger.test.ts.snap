// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger > should set custom logger > cannot set log level when using custom logger 1`] = `[BearAuthError: Cannot set log level when using custom logger.]`;
